pandas/io/formats/excel.py

```diff
@@ -722,39 +716,40 @@ def _format_hierarchical_rows(self):
                     )
 
                     for i in spans:
-                        if spans[i] > 1:
-                            yield ExcelCell(
-                                self.rowcounter + i,
-                                gcolidx,
-                                values[i],
-                                self.header_style,
-                                self.rowcounter + i + spans[i] - 1,
-                                gcolidx,
-                            )
-                        else:
-                            yield ExcelCell(
-                                self.rowcounter + i,
-                                gcolidx,
-                                values[i],
-                                self.header_style,
-                            )
+                        yield ExcelCell(
+                            row=self.rowcounter + i,
+                            col=gcolidx,
+                            val=values[i],
+                            style=self.header_style,
+                            mergestart=(
+                                self.rowcounter + i + spans[i] - 1
+                                if spans[i] > 1
+                                else None
+                            ),
+                            mergeend=gcolidx if spans[i] > 1 else None,
+                        )
                     gcolidx += 1
 
             else:
                 # Format hierarchical rows with non-merged values.
                 for indexcolvals in zip(*self.df.index):
                     for idx, indexcolval in enumerate(indexcolvals):
                         yield ExcelCell(
-                            self.rowcounter + idx,
-                            gcolidx,
-                            indexcolval,
-                            self.header_style,
+                            row=self.rowcounter + idx,
+                            col=gcolidx,
+                            val=indexcolval,
+                            style=self.header_style,
                         )
                     gcolidx += 1
 
         yield from self._generate_body(gcolidx)
 
-    def _generate_body(self, coloffset: int):
+    @property
+    def _has_aliases(self) -> bool:
+        """Whether the aliases for column names are present."""
+        return not isinstance(self.header, bool)
```

Comment:
thinking some more, maybe best not to use bool.

we've never really had a discussion on use of bool in type annotations, but sometimes we are actually accepting any truthy value and other times we explicitly check if x is False or if x is True.

the header parameter of the docstring of ExcelFormatter should probably be updated to match. It currently is  
```
    header : boolean or list of string, default True
        Write out column names. If a list of string is given it is
        assumed to be aliases for the column names
```
even though (tuple, list, np.ndarray, ABCIndex) is the current accepted types on master.

and the current mypy issue is that header is annotated as  `Union[Sequence[Label], bool]`

perhaps use is_list_like here. (ignoring that a str is a Sequence)

Reply:
I replaced with ``is_list_like``.
Also slightly updated docstring. Will it work?

Reply:
there is still a discrepancy. Can a sequence of any hashable be passed or does it have to be a sequence of strings.

Reply:
``ExcelFormatter`` is called only from ``Styler.to_excel`` and ``NDFrame.to_excel``.
According to these docstrings
```
header : bool or list of str, default True
```
I also looked through the test base and noticed that header is either bool or some sequence of strings (on some occasions).

I am not aware of other use cases, when header is a sequence of any hashables rather than strings.

Reply:
ok can leave as is for now.

The docstrings of public methods document the public api, the type annotations should normally be as permissible as possible for argument positions and the docstring and type annotations should match.

we don't want to change the public api without discussion, so can circle back to this.